Shipping Labels: Create/activate new package on Add New Package screen

rachelmcr opened the pull request:
Part of: #4745

## Description

This makes the Done button functional on the Add New Package screen in the shipping labels flow.

When the Done button is tapped, the new custom or service package is created/activated remotely and the screen is dismissed. The new package is selected for the shipping label and is listed in the Package Selected screen (the screen before the Add New Package screen if the store has existing packages).

Note: If the API calls fail, the Add New Package screen is not dismissed. However, this PR does not include any user-facing error messaging. We should show a notice if there is an error, but it may take some work to get the notice presenter working with SwiftUI so I haven't included that here.

## Changes

Networking:

* Updates the `createPackage` endpoint in `ShippingLabelRemote` to support activating a service package.
* Updates the models for `ShippingLabelPackagesResponse` and `ShippingLabelPredefinedOption` to save the provider ID (e.g. "usps" or "dhlexpress") for predefined options. This is used by the `createPackage` endpoint when activating a service package.
* Updates the Fakes and unit tests with those changes.

Yosemite:

* Updates the Shipping Label Store and Action to support the changes to the `createPackage` endpoint.

UI:

* Makes the Done button functional in `ShippingLabelAddNewPackage`:
   * Uses the state property `isSyncing` to display an activity indicator and disable the done button while the API calls are in progress.
   * If the API calls are successful, it dismisses the view.
* Adds helpers to `ShippingLabelAddNewPackageViewModel` for the Done button actions (`createCustomPackage` and `activateServicePackage`). It gets the new custom/service package from the child view, creates/activates the package remotely, syncs the package list from remote, and passes the new package and updated package list back to the Package List / Package Details.

<img src="https://user-images.githubusercontent.com/8658164/132456658-1f7f9aeb-1383-4716-afeb-5850b522a18e.gif" width="300px">


## Testing

1. Make sure you have configured the WooCommerce WCShip plugin.
2. Launch the app in debug mode (so the feature flag is enabled).
3. Open an order detail that is eligible for the shipping label.
4. Tap the button "Create Shipping Label.”
5. Confirm "Ship From" and "Ship To.”
6. In the package details section, tap on “Package Selected.”
7. If you already have existing packages on your store, you will see the Package Selected screen. Tap on "Create new package" at the bottom of the screen.
8. On the Add New Package screen, fill in the Custom Package form or select a Service Package.
9. Tap the Done button and notice there is an activity indicator and the button is disabled, and when it finishes you are taken back to the previous screen where the new package is selected.

## Submitter Checklist

Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

rachelmcr committed:
Add support for activating a service package for shipping labels

rachelmcr committed:
Update packages endpoint to support activating both custom and service packages if both are provided

rachelmcr committed:
Update shipping label tests to include support for predefined option in packages endpoint

rachelmcr committed:
Fix createPackages parameters when no custom package is included

rachelmcr committed:
Add unit tests for different combinations of packages sent to createPackage endpoint

rachelmcr committed:
Handle new package creation and selection

rachelmcr committed:
Reset child view state after dismissing view

rachelmcr committed:
Only display activity indicator when custom package is validated

rachelmcr committed:
Clean up unused property

rachelmcr committed:
Simplify API calls and dismiss action

rachelmcr committed:
Add comments

rachelmcr committed:
Fix hound violation

rachelmcr committed:
Explicitly capture weak self

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Package Details/Package Selection/Package Creation/ShippingLabelAddNewPackageViewModel.swift:
```diff
@@ -5,23 +5,157 @@ import Yosemite
     // View models for child views (tabs)
-    let customPackageVM = ShippingLabelCustomPackageFormViewModel()
-    let servicePackageVM: ShippingLabelServicePackageListViewModel
+    var customPackageVM = ShippingLabelCustomPackageFormViewModel()
+    lazy var servicePackageVM = ShippingLabelServicePackageListViewModel(packagesResponse: packagesResponse)
```
Nit: I think these should be `private(set)` to make sure that the variables can only be updated within this view model.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Package Details/Package Selection/Package Creation/ShippingLabelAddNewPackageViewModel.swift:
```diff
@@ -5,23 +5,157 @@ import Yosemite
+            // On success, reset tab state and save new package details
+            guard let self = self, success else { return }
+            self.customPackageVM = ShippingLabelCustomPackageFormViewModel()
+            self.servicePackageVM.packagesResponse = self.packagesResponse
```
I'd recommend avoiding direct mutations of objects from outside. For instance, since the service package view model's package response is internal and mutable, there's a risk that it can be updated accidentally from outside. This can be updated with better access control as follows:
- Keep the service package view model's `packagesResponse` private and immutable.
- Make `servicePackageVM` a variable instead of a lazy one so that you can reset it with a new package response when needed.

rachelmcr replied:
Thank you! That's good advice for me to keep in mind, and makes a lot of sense here.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Package Details/ShippingLabelPackageDetailsViewModel.swift:
```diff
@@ -83,7 +83,11 @@ final class ShippingLabelPackageDetailsViewModel: ObservableObject {
+    lazy var addNewPackageViewModel = ShippingLabelAddNewPackageViewModel(siteID: order.siteID,
+                                                                          packagesResponse: packagesResponse,
+                                                                          onCompletion: { (customPackage, predefinedOption, packagesResponse) in
+                                                                            self.handleNewPackage(customPackage, predefinedOption, packagesResponse)
```
Since `onCompletion` is an `@escaping` block, it's safer to use capture `self` weakly here.

itsmeichigo reviewed (changes requested):
I have done a quick test and both adding custom packages and predefined packages work as expected 💯

However it seems that the package response is not synced on the purchase form yet. I think the form should trigger syncing package response on `viewWillAppear`? Or we can use a callback if that's preferred.

<img src="https://user-images.githubusercontent.com/5533851/132650355-5ff780dd-d7a3-4d40-bdaa-6370d1d45981.gif" width=400 />

Plus - do you think it's necessary to add unit tests to the add package view model to check whether view models are reset after successfully creating packages?

I have also left some comments regarding access control and memory management.

rachelmcr committed:
Improve access control for view model properties

rachelmcr committed:
Capture self weakly for completion block

rachelmcr committed:
Handle changes to package response on shipping label purchase form

rachelmcr committed:
Add unit tests for Add New Package view model

rachelmcr committed:
Merge branch 'develop' into issue/4745-service-packages-endpoint

rachelmcr commented:
> However it seems that the package response is not synced on the purchase form yet. I think the form should trigger syncing package response on viewWillAppear? Or we can use a callback if that's preferred.

Oh, great catch, thank you! I decided to use a callback since we're already syncing the package response after the new package is added on the add new package screen, so we can pass it back to the purchase form and handle it along with other changes to package details. (Also, that way we don't sync the package response every time the purchase form appears.)

> Plus - do you think it's necessary to add unit tests to the add package view model to check whether view models are reset after successfully creating packages?

Oh yes! I should have mentioned that I was planning to follow up with more unit tests, but I just finished up tests for that so I went ahead and added them to this PR.

@itsmeichigo I think I've addressed all of your feedback and resolved the merge conflicts, so this is ready for another look!

itsmeichigo commented on WooCommerce/WooCommerceTests/ViewRelated/Shipping Label/Create Shipping Label/ShippingLabelAddNewPackageViewModelTests.swift:
```diff
@@ -0,0 +1,89 @@
+        viewModel.createCustomPackage(onCompletion: { _ in })
+
+        // Then
+        XCTAssertNil(viewModel.customPackageVM.validatedCustomPackage)
```
Suggestion: it'd be nice to also assert that all fields are reset to default values?

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Package Details/ShippingLabelPackageDetails.swift:
```diff
@@ -8,7 +8,7 @@ struct ShippingLabelPackageDetails: View {
     /// Completion callback
     ///
-    typealias Completion = (_ selectedPackages: [ShippingLabelPackageAttributes]) -> Void
+    typealias Completion = (_ packagesResponse: ShippingLabelPackagesResponse?, _ selectedPackages: [ShippingLabelPackageAttributes]) -> Void
```
I think just updating package response on completion (after selection a package) is not sufficient. There's a case where a user adds new package but then navigates out of the Package Details screen without selecting any package - here the package response is not updated. 

I'd suggest creating a new callback block just for updating package response, inject into the view model and call it at the end of `handleNewPackage` method. That way we can be sure that the package list is updated whenever a new package is added.

itsmeichigo replied:
It's also nice if we can add both blocks to the view model's initializer instead of keeping one in the view. Tapping Done will then trigger a method in the view model that calls the completion block.

rachelmcr replied:
Good catch! I didn't consider that case before. Thank you for the really clear suggestions; that makes a lot of sense. I made those suggested changes in 08a9e47, including putting both blocks in the view model.

itsmeichigo reviewed (changes requested):
Thanks for updating the unit tests! I found that there's still a case where package response is not updated on the purchase form. Once that's resolved I think we'll be good to go!

rachelmcr committed:
Ensure package response is updated separately from saving new package details

rachelmcr committed:
Assert that all fields are reset to default values

rachelmcr commented:
Ok @itsmeichigo, this is ready for another look! In particular, now it should be able to handle updating the package response on the purchase form even when you don't save changes to the package details.

itsmeichigo reviewed (approved):
Everything looks great now! Thank you for your patience! :shipit:
